ot/gromov/_unbalanced.py

```diff
@@ -0,0 +1,1210 @@
+            if eps_samp != 0:
+                div_samp = div(pi_samp, pi1_samp, pi2_samp, px_samp, py_samp)
+                ucoot_cost = ucoot_cost + eps_samp * div_samp
+            if eps_feat != 0:
+                div_feat = div(pi_feat, pi1_feat, pi2_feat, px_feat, py_feat)
+                ucoot_cost = ucoot_cost + eps_feat * div_feat
+
+        return linear_cost, ucoot_cost
+
+    # Support functions for squared L2 norm
+    def parameters_uot_l2(pi, tuple_weights, hyperparams):
+        """Compute parameters of the L2 loss."""
+
+        rho_x, rho_y, eps = hyperparams
+        wx, wy, wxy = tuple_weights
+
+        pi1, pi2 = nx.sum(pi, 1), nx.sum(pi, 0)
+        l2_pi1, l2_pi2, l2_pi = nx.sum(pi1**2), nx.sum(pi2**2), nx.sum(pi**2)
+
+        weighted_wx = wx * nx.sum(pi1 * wx) / l2_pi1
+        weighted_wy = wy * nx.sum(pi2 * wy) / l2_pi2
+        weighted_wxy = wxy * nx.sum(pi * wxy) / l2_pi if reg_type == "joint" else wxy
+        weighted_w = (weighted_wx, weighted_wy, weighted_wxy)
+
+        new_rho = (rho_x * l2_pi1, rho_y * l2_pi2)
+        new_eps = eps * l2_pi if reg_type == "joint" else eps
+
+        return weighted_w, new_rho, new_eps
+
+    # UOT solver for KL and squared L2
+    def uot_solver(wx, wy, wxy, cost, eps, rho, init_pi, init_duals):
+        if unbalanced_solver == "scaling":
+            pi, log = sinkhorn_unbalanced(
+                a=wx, b=wy, M=cost, reg=eps, reg_m=rho, reg_type="kl",
+                warmstart=init_duals, method=method_sinkhorn,
+                numItermax=max_iter_ot, stopThr=tol_ot, verbose=False, log=True)
+            duals = (log['logu'], log['logv'])
+
+        elif unbalanced_solver == "mm":
+            pi = mm_unbalanced(a=wx, b=wy, M=cost, reg_m=rho,
+                               c=wxy, reg=eps, div=divergence,
+                               G0=init_pi, numItermax=max_iter_ot,
+                               stopThr=tol_ot, verbose=False, log=False)
+            duals = (None, None)
+
+        elif unbalanced_solver == "lbfgsb":
+            pi = lbfgsb_unbalanced(a=wx, b=wy, M=cost, reg=eps, reg_m=rho,
+                                   c=wxy, reg_div=divergence,
+                                   regm_div=divergence,
+                                   G0=init_pi, numItermax=max_iter_ot,
+                                   stopThr=tol_ot, method='L-BFGS-B',
+                                   verbose=False, log=False)
+            duals = (None, None)
+
+        return pi, duals
+
+    # MAIN FUNCTION
+
+    if reg_type not in ["joint", "independent"]:
+        raise (NotImplementedError('Unknown reg_type="{}"'.format(reg_type)))
+    if divergence not in ["kl", "l2"]:
+        raise (NotImplementedError('Unknown divergence="{}"'.format(divergence)))
+    if unbalanced_solver not in ["scaling", "mm", "lbfgsb"]:
+        raise (NotImplementedError('Unknown method="{}"'.format(unbalanced_solver)))
+
+    X, Y = list_to_array(X, Y)
+    nx = get_backend(X, Y)
```

Comment:
you need to check the backends for all potential inputs. Can create a list arr = [X, Y], where you append other inputs e.g if there are not set to None, initial transport plans, marginals etc...

Reply:
Fixed now